content/advent-2015/glow-map-reduce-for-golang.md

```diff
@@ -0,0 +1,485 @@
+mapper's result emitter. Usually 1 mapper emits 1 result, which
+can be represented as function's return result, e.g.,
+``` func(key string) int ```. But if a mapper can emit 0 or more
+than 1 result, this won't work. The channel fits nicely here. If the
+last parameter of a mapper funtion is a channel, e.g.,
+``` func(line string, ch chan string) ```, it will be treated
+as the result emitter.
+
+# Setup Glow Cluster
+
+Now let us setup the Glow cluster, just in case you need to scale up.
+Setting up Glow Cluster is super easy. First, build the "glow" binary file:
+
+```
+$ go get github.com/chrislusf/glow
+```
+
+The compiled binary file is usually ```$GOPATH/bin/glow```.
+Now copy it to any computer, and run this:
+```
+// start a master on one computer
+> glow master
+// run one or more agents on computers
+> glow agent --dir ./glow/data --max.executors=16 --memory=2048 --master="localhost:8930" --port 8931
+// it's fine to run several agents on the same computer
+> glow agent --dir ./glow/data2 --max.executors=8 --memory=512 --master="localhost:8930" --port 8932
+```
+
+Either master or agent only takes 5~6 MB memory. They are quiet
+and efficient. I highly recommend run this "glow agent" on any
+machine you can find, so that you can tap into the compute power any
+time you want, with one line of code change as follows.
+
+# Distributed Execution
+
+Just insert this line to the import section, around line 7:
+```go
+    _ "github.com/chrislusf/glow/driver"
+```
+
+
+Now the word count can run distributedly this way:
+```
+$ go run t.go -glow
+2015/12/08 00:54:46 localhost:8930 allocated 1 executors.
+2015/12/08 00:54:46 localhost:8930 allocated 1 executors.
+2015/12/08 00:54:46 localhost:8930 allocated 2 executors.
+2015/12/08 00:54:46 localhost:8930 allocated 1 executors.
+127.0.0.1:8931>count: 532
+```
+
+
+Also, you can visualize the flow. Run this command:
+```
+$ go run t.go -glow -glow.flow.plot > x.dot
+$ dot -Tpng -owc.png x.dot
+```
+
+And the flow graph looks like this:
+
+![Word Count Flow](/postimages/advent-2015/glow-word-count-flow.png)
+
+## Something wrong!
+
+You may get a different result when running distributedly vs
+standalone mode! This is because I cheated!
+
+The input file "/etc/passwd" could be different on different servers.
+The file should be sent to the executor that's running the
+TextFile() function. Let's change the TextFile() call to this:
+```go
+  TextFile("passwd", 3)
+```
+And make sure the file is shipped to the executor:
+
+```
+$ go run t.go -glow -glow.related.files="/etc/passwd"
+...
+127.0.0.1:8931>count: 532
+```
+
+As you can see, still many things can break when it comes to
+distributed mode. We need to understand how Glow works first.
+
+# Glow Underneath
+## How Glow works?
+The word count code looks simple, but does many things. It can act as either a driver or an executor depending on command line
+parameters.
+
+When the driver starts, it will ask Glow master for resources. Glow
```

Comment:
I think "Glow master" needs a "the" in the beginning, so "will ask the Glow master" ... in a number of occasions below as well, where the Glow master is referred to.